File: omv-client/src/entities/Rule.js
import { OwlEntity } from "./OwlEntity.js";

export class Rule {
    /** @type {String} */ label = "";
    /** @type {String} */ relation = "";
    /** @type {String} */ direction = "";
    /** @type {String} */ comments = "";
    /** @type {String} */ similarity = "";
    /** @type {String} */ simcomments = "";
    /** @type {Object} */ directTransformation = null;
    /** @type {Object} */ inverseTransformation = null;
    /** @type {Object} */ entity1 = null;
    /** @type {Object} */ entity2 = null;

    /** @param {Object} ruleobject
     * @returns {Rule}
     */
    static fromObject(ruleobject) {
        let rule = new Rule();
        rule.label = ruleobject.label;
        rule.relation = ruleobject.relation;
        rule.direction = ruleobject.direction;
        rule.comments = ruleobject.comments;
        rule.similarity = ruleobject.similarity;
        rule.simcomments = ruleobject.simcomments;
        rule.directTransformation = ruleobject.directTransformation;
        rule.inverseTransformation = ruleobject.inverseTransformation;
        rule.entity1 = ruleobject.entity1;
        rule.entity2 = ruleobject.entity2;
        return rule;
    }

    /** @param {Object[]} rulesobject
     * @return {Rule[]}
    */
    static listFromObject(rulesobject) {
        let rules = [];
        for (let ruleobject of rulesobject) {
            rules.push(Rule.fromObject(ruleobject));
        }
        return rules;
    }

    /** @param {Rule[]} rules
     * @param {Rule} rule
     * @return {Boolean}
    */
    static listExistsRule(rules, rule) {
        for (let rl of rules) {
            if (rl.label == rule.label) {
                return true;
            }
        }
        return false;
    }

    /** @param {Rule[]} rules
     * @param {OwlEntity} owlclasses
     * @return {Rule[]}
    */
    static findRule(rules, owlNumber, entityType, owlentity, orderedRules) {
        for (let rule of rules) {
            let ruleowl = (owlNumber == 1) ? rule.entity1 : rule.entity2;
            let ruleentities = [];
            if (entityType == "class") {
                ruleentities = ruleowl.classes;
            } else if (entityType == "objprop") {
                ruleentities = ruleowl.objectprops;
            } else {
                ruleentities = ruleowl.dataprops;
            }
            for (let ruleentity of ruleentities) {
                if (owlentity.iri == ruleentity.iri) {
                    if (!Rule.listExistsRule(orderedRules, rule)) {
                        orderedRules.push(rule);
                    }
                }
            }
        }
        for (let child of owlentity.children) {
            Rule.findRule(rules, owlNumber, entityType, child, orderedRules);
        }
    }

    /** @param {Rule[]} rules
     * @param {OwlEntity} owlclasses
     * @return {Rule[]}
    */
    static listOrdered(rules, owlNumber, owlclasses, objprops, dataprops) {
        let rulesclone = Object.assign({}, rules);
        let orderedRules = [];
        Rule.findRule(rules, owlNumber, "class", owlclasses, orderedRules);
        return orderedRules;
    }
}

File: omv-server/src/main/java/omv/server/entities/Rule.java
package omv.server.entities;

import java.util.ArrayList;

import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;

public class Rule {
    public String label = "";
    public String relation = "";
    public String direction = "";
    public String comments = "";
    public String similarity = "";
    public String simcomments = "";
    public JsonObject directTransformation = new JsonObject();
    public JsonObject inverseTransformation = new JsonObject();
    public RuleEntity entity1 = new RuleEntity();
    public RuleEntity entity2 = new RuleEntity();

    public void init(JsonObject ruleobject) {
        this.relation = ruleobject.getString("relation");
        this.direction = ruleobject.getString("direction");
        this.comments = ruleobject.getString("comments");
        this.similarity = ruleobject.getString("similarity");
        this.simcomments = ruleobject.getString("simcomments");
        this.directTransformation = ruleobject.getJsonObject("directTransformation");
        this.inverseTransformation = ruleobject.getJsonObject("inverseTransformation");
        JsonObject entity1object = ruleobject.getJsonObject("entity1");
        this.entity1.init(entity1object);
        JsonObject entity2object = ruleobject.getJsonObject("entity2");
        this.entity2.init(entity2object);
    }

    public JsonObject toJsonObject() {
        JsonObject rulejsonobject = new JsonObject();
        rulejsonobject.put("label", this.label);
        rulejsonobject.put("relation", this.relation);
        rulejsonobject.put("direction", this.direction);
        rulejsonobject.put("comments", this.comments);
        rulejsonobject.put("similarity", this.similarity);
        rulejsonobject.put("simcomments", this.simcomments);
        rulejsonobject.put("directTransformation", this.directTransformation);
        rulejsonobject.put("inverseTransformation", this.inverseTransformation);
        rulejsonobject.put("entity1", this.entity1.toJsonObject());
        rulejsonobject.put("entity2", this.entity2.toJsonObject());
        return rulejsonobject;
    }

    public static JsonArray listToJsonArray(ArrayList<Rule> rules) {
        JsonArray rulesjsonobject = new JsonArray();
        rules.forEach((rule) -> {
            rulesjsonobject.add(rule.toJsonObject());
        });
        return rulesjsonobject;

    }
}

class RuleEntity {
    public String pid;
    public JsonArray classes;
    public JsonArray objectprops;
    public JsonArray dataprops;

    public RuleEntity() {
        this.pid = "";
        this.classes = new JsonArray();
        this.objectprops = new JsonArray();
        this.dataprops = new JsonArray();
    }

    public void init(JsonObject entityobject) {
        this.pid = entityobject.getString("pid");
        this.determineAndAddOwlEntity(entityobject);

    }

    private void determineAndAddOwlEntity(JsonObject obj) {
        if (obj.getString("classuri") != null) {
            this.addOwlEntity(obj, "class");
        }
        if (obj.getString("relationuri") != null) {
            this.addOwlEntity(obj, "objectprop");
        }
        if (obj.getString("propertyuri") != null) {
            this.addOwlEntity(obj, "dataprop");
        }
        // detect folded rules
        if (obj.getJsonObject("property") != null) {
            this.determineAndAddOwlEntity(obj.getJsonObject("property"));
        }
        if (obj.getJsonObject("relation") != null) {
            this.determineAndAddOwlEntity(obj.getJsonObject("relation"));
        }
        if (obj.getJsonObject("domainclass") != null) {
            this.determineAndAddOwlEntity(obj.getJsonObject("domainclass"));
        }
        if (obj.getJsonObject("rangeclass") != null) {
            this.determineAndAddOwlEntity(obj.getJsonObject("rangeclass"));
        }
        //
        JsonObject cls = obj.getJsonObject("cls");
        if (cls != null) {
            this.determineAndAddOwlEntity(cls);
        }
        JsonArray classarray = obj.getJsonArray("classarray");
        if (classarray != null) {
            classarray.forEach((obj2) -> {
                this.determineAndAddOwlEntity((JsonObject) obj2);
            });;
        }
        JsonArray proparray = obj.getJsonArray("proparray");
        if (proparray != null) {
            proparray.forEach((obj2) -> {
                this.determineAndAddOwlEntity((JsonObject) obj2);
            });;
        }
        JsonArray relationarray = obj.getJsonArray("relationarray");
        if (relationarray != null) {
            relationarray.forEach((obj2) -> {
                this.determineAndAddOwlEntity((JsonObject) obj2);
            });;
        }
    }

    private void addOwlEntity(JsonObject obj, String type) {
        JsonObject newowlentity = new JsonObject();
        newowlentity.put("pid", obj.getString("pid"));
        newowlentity.put("index", obj.getInteger("index"));
        if (type.equals("class")) {
            newowlentity.put("iri", obj.getString("classuri"));
            this.classes.add(newowlentity);
        } else if (type.equals("objectprop")) {
            newowlentity.put("iri", obj.getString("relationuri"));
            this.objectprops.add(newowlentity);
        } else if (type.equals("dataprop")) {
            newowlentity.put("iri", obj.getString("propertyuri"));
            this.dataprops.add(newowlentity);
        }
    }

    public JsonObject toJsonObject() {
        JsonObject entityjsonobject = new JsonObject();
        entityjsonobject.put("pid", this.pid);
        entityjsonobject.put("classes", this.classes);
        entityjsonobject.put("objectprops", this.objectprops);
        entityjsonobject.put("dataprops", this.dataprops);
        return entityjsonobject;
    }
}